src/resources/rmt_esp32.cc

```diff
@@ -17,459 +17,525 @@
 
 #ifdef TOIT_ESP32
 
-#include <driver/rmt.h>
+#include <soc/soc.h>
+#include <driver/rmt_tx.h>
+#include <driver/rmt_rx.h>
 #include <driver/gpio.h>
-#include <hal/gpio_hal.h>
 
 #include "../objects_inline.h"
 #include "../primitive.h"
 #include "../process.h"
 #include "../resource.h"
-#include "../resource_pool.h"
+
+
+#include "../event_sources/system_esp32.h"
+#include "../event_sources/ev_queue_esp32.h"
+
+#if CONFIG_RMT_ISR_IRAM_SAFE || CONFIG_RMT_RECV_FUNC_IN_IRAM
+#define RMT_MEM_ALLOC_CAPS      (MALLOC_CAP_INTERNAL | MALLOC_CAP_8BIT)
+#define RMT_IRAM_ATTR IRAM_ATTR
+#else
+#define RMT_MEM_ALLOC_CAPS      MALLOC_CAP_DEFAULT
+#define RMT_IRAM_ATTR
+#endif
 
 namespace toit {
 
-const rmt_channel_t kInvalidChannel = static_cast<rmt_channel_t>(-1);
+const int kReadState = 1 << 0;
+const int kWriteState = 1 << 1;
 
-// Copied from rmt_legacy.c.
-#define RMT_RX_CHANNEL_ENCODING_START (SOC_RMT_CHANNELS_PER_GROUP-SOC_RMT_TX_CANDIDATES_PER_GROUP)
-#define RMT_TX_CHANNEL_ENCODING_END   (SOC_RMT_TX_CANDIDATES_PER_GROUP-1)
+typedef struct Event {
+  word is_read : 1;
+  word read_symbols_count : 31;
+} Event;
 
-// Copied from rmt_legacy.c.
-#define RMT_IS_RX_CHANNEL(channel) ((channel) >= RMT_RX_CHANNEL_ENCODING_START)
-#define RMT_IS_TX_CHANNEL(channel) ((channel) <= RMT_TX_CHANNEL_ENCODING_END)
+class RmtResourceGroup : public ResourceGroup {
+ public:
+  TAG(RmtResourceGroup);
+  RmtResourceGroup(Process* process, EventSource* event_source)
+    : ResourceGroup(process, event_source) {}
 
-#ifndef SOC_RMT_CHANNELS_PER_GROUP
-#error "SOC_RMT_CHANNELS_PER_GROUP not defined"
-#endif
+  uint32_t on_event(Resource* r, word data, uint32_t state) override {
+    if (data == kReadState || data == kWriteState) {
+      state |= data;
+    }
+    return state;
+  }
+};
 
-#if SOC_RMT_CHANNELS_PER_GROUP == SOC_RMT_TX_CANDIDATES_PER_GROUP
-// All channels are TX channels. We assume that they also are RX channels.
-// This is the setup on the ESP32 and ESP32S2 chips.
-static ResourcePool<rmt_channel_t, kInvalidChannel> rmt_channels_(
-    RMT_CHANNEL_0, RMT_CHANNEL_1
-#if SOC_RMT_CHANNELS_PER_GROUP > 2
-    , RMT_CHANNEL_2, RMT_CHANNEL_3
-#if SOC_RMT_CHANNELS_PER_GROUP > 4
-    , RMT_CHANNEL_4, RMT_CHANNEL_5, RMT_CHANNEL_6, RMT_CHANNEL_7
-#endif
-#endif
-);
+class RmtResource;
 
-static ResourcePool<rmt_channel_t, kInvalidChannel>* rmt_tx_channels = &rmt_channels_;
-static ResourcePool<rmt_channel_t, kInvalidChannel>* rmt_rx_channels = &rmt_channels_;
+class RmtInOut {
+ public:
+  virtual ~RmtInOut() {}
+  virtual esp_err_t disable() = 0;
+};
 
-#else
-// Some channels are TX, some are RX.
-// This is the configuration on the ESP32C3, ESP32H2, and ESP32S3 chips.
-// The TX channels are always before the RX channels.
+class RmtIn : public RmtInOut {
```

Comment:
Currently the differences between RmtIn and RmtOut aren't that big, but there will be more changes.